### PR TITLE
Update EnergyDispersion

### PR DESCRIPTION
**Pull request by joleroi**
This adresses https://github.com/gammapy/gammapy/issues/953

* [x] Change interpolation behaviour of ``EnergyDispersion`` to nearest neighbour. Since the EnergyDispersion matrix contains integral probabilities interpolation does not make sense (for coarse binning). The ``EnergyDispersion`` class is only a wrapper around a fix ``np.array``. I decided to still use ``NDDataArray`` to keep the axis informations (and because the change was easier that way)

* [x] Rewrite ``calculated_prediced_counts`` as class. No interpolation is performed at this steps. The inputs need to have the desired binning

Adresses https://github.com/gammapy/gammapy/issues/953 and https://github.com/gammapy/gammapy/issues/882

### CODE REVIEW

File: gammapy/data/tests/test_data_store.py [hunk 1]
```diff
@@ -195,6 +195,6 @@ def test_make_mean_edisp(tmpdir):
     i = np.where(rmf.data.evaluate(e_reco=Energy(1.5, "TeV")) != 0)[0]
     i2 = np.where(rmf2.data.evaluate(e_reco=Energy(1.5, "TeV")) != 0)[0]
     assert_equal(i, i2)
-    i = np.where(rmf.data.evaluate(e_reco=Energy(55, "TeV")) != 0)[0]
-    i2 = np.where(rmf2.data.evaluate(e_reco=Energy(55, "TeV")) != 0)[0]
+    i = np.where(rmf.data.evaluate(e_reco=Energy(40, "TeV")) != 0)[0]
```
Comment (joleroi) [inline gammapy/data/tests/test_data_store.py]: @leajouvin 

Can you check if this change does not break the test?

**Comment by joleroi** [comment]
The lightcurve functionality is broken by this PR, because the predicted counts calculation does not accept ``e_reco`` anymore.
https://github.com/joleroi/gammapy/blob/fix_edisp/gammapy/time/lightcurve.py#L324
@jjlk Did you add this? Can you fix it, I don't quickly understand what the desired behaviour is.

**Comment by joleroi** [comment]
> @jjlk Did you add this? Can you fix it, I don't quickly understand what the desired behaviour is.

I sat down with @cdeil and we manage to recover the functionality using minimal changes.

**Comment by joleroi** [comment]
I will fix the failing notebook tests locally. Merging!

**Comment by jjlk** [comment]
Hi @joleroi , @cdeil, 
Question, why the full-containment correction is broken? I see that you changed the test for the light curves and remove the full-containment correction: https://github.com/gammapy/gammapy/pull/956/commits/e2216273d2bf1a2e5a9812e1ab5146dc01de2779. Is there a reason for that? We need this functionality, right? Especially for the light curves that we have to produce for the HESS data release. 

Now, for the light curves we got NaN values for the expected excess when we use the full-containment correction.

What is the status? Thanks ++

**Comment by joleroi** [comment]
Hi, The containment correction adds NaN values to the effective area because of this line
https://github.com/gammapy/gammapy/blob/master/gammapy/spectrum/extract.py#L231
I talked to @registerrier (who added the containment correction) and he said he put it to have something, so it could be improved in principle. It seem like the NaN case occures if the PSF integration fails. I don't have a good idea how to change this.

**Comment by jjlk** [comment]
Ola @joleroi , 
Okay. Do you have any idea of which part of the changes that have been made five days ago broke the computation of the correction?

**Comment by joleroi** [comment]
Yes, before this function was called 
https://github.com/gammapy/gammapy/pull/956/files#diff-293aaf2e09b165c538046fd598fe125bL162
which was introduces to fill nan values at the low energy and high energy part of the effective area (the range outside the interpolator). This hid the 'bug' in the containment correction calculation.

**Pull request merged**